runtime/interpreter/value.go

```diff
@@ -17374,6 +17435,15 @@ func (v *StorageReferenceValue) GetTypeKey(
 ) Value {
 	self := v.mustReferencedValue(interpreter, locationRange)
 
+	if selfComposite, isComposite := self.(*CompositeValue); isComposite {
```

Comment:
Why is this special case for `*CompositeValue` needed?

Reply:
The implementation of `GetTypeKey` for `CompositeValue`s is written under the assumption that it is called with an owned value, and calls into `getTypeKey` with the full domain of the attachment's entitlement map as the attachment's access. This is not the behavior we want when we are accessing an attachment on a reference to a composite, so we compute the entitlement set of the reference and call `getTypeKey` with that access instead.